### PR TITLE
Migrate to react-navigation v5

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

- Dogfood our libraries
- Required for #8064 
- We use a couple hacks to make bare-expo (E2E) linking work in v4, this should fix those. 

# How

- Convert linking format in bare-expo to `test-suite/run?tests=Basic%20Constants` from `test-suite/select/Basic,Constants`
- Replace static navigation with dynamic v5 navigation
- Convert screens using `NavigationEvents` to functional components and replace with `useFocusEffect` hook
  - Implement `use-expo/permissions` in place of some focus events
- Add masked view to `bare-expo`
- Delete type shims for `react-navigation-stack`
- Migrate to `react-navigation-shared-element@prerelease`
- Remove `react-navigation` `react-navigation-stack`, `@react-navigation/web`, `react-navigation-tabs`

## Standalone changes

Some features can be moved to other PRs for brevity.

- A few screen rewrites can be partially moved
- masked-view installation https://github.com/expo/expo/pull/8858
- NCL RNW 11 component warnings being muted
- bare-expo breaks on missing slider package, we can add that elsewhere

# Test Plan

- Test suite should be usable from the (latest) client
  - [x] iOS
  - [x] Android
  - [x] web
- NCL should be able to open all of the same screens as before
  - [x] iOS 
  - [x] Android
  - [x] web
- `bare-expo` should be able to load test-suite and the expo screens from NCL
  - [x] iOS
  - [x] Android
  - [x] web
- Linking should work in `bare-expo` 
  - This should work for iOS and Android: `npx uri-scheme open bareexpo://test-suite/run?tests=Basic`
- `bare-expo` E2E tests should continue to pass
- Navigation should be styled the same as before

### CODE REVIEW

File: apps/native-component-list/package.json [hunk 1]
```diff
@@ -89,7 +92,7 @@
     "react-native-gesture-handler": "~1.6.0",
     "react-native-is-iphonex": "^1.0.1",
     "react-native-maps": "0.27.1",
-    "react-native-paper": "github:brentvatne/react-native-paper#@brent/fix-bottom-navigation-rn-57",
```
Comment (EvanBacon) [inline apps/native-component-list/package.json:92]: What does this do? Will it conflict with V5?

Comment (brentvatne) [reply apps/native-component-list/package.json:92]: it looks like this was just a fix for https://github.com/react-navigation/react-navigation/issues/7466 and we should be able to update react-native-paper to latest

File: apps/bare-expo/MainNavigator.tsx [hunk 2]
```diff
@@ -53,40 +43,76 @@ const Redirect = optionalRequire(() =>
...
-      left: 'always',
+const Tab = createBottomTabNavigator();
+
+const Switch = createStackNavigator();
+
+const linking = {
+  prefixes: [Platform.select({ web: Linking.makeUrl('/'), default: 'bareexpo://' })],
+  config: {
+    main: {
+      path: '',
```
Comment (satya164) [inline apps/bare-expo/MainNavigator.tsx:58]: There was slight change in this config so would be good to upgrade to latest versions of navigation to use those  :)

then it'll be like:

```js
config: {
  screens: {
    main: ...
  }
}
```

File: apps/bare-expo/MainNavigator.tsx [hunk 3]
```diff
@@ -53,40 +43,76 @@ const Redirect = optionalRequire(() =>
...
+      initialRouteName: 'test-suite',
+      screens: {
+        'test-suite': {
+          path: 'test-suite',
+          screens: {
+            select: '',
+            run: '/run',
```
Comment (satya164) [inline apps/bare-expo/MainNavigator.tsx:65]: The leading `slash` is not needed

File: apps/bare-expo/MainNavigator.tsx [hunk 4]
```diff
@@ -53,40 +43,76 @@ const Redirect = optionalRequire(() =>
...
-if (Redirect) {
-  switchRoutes.redirect = Redirect;
+function TabNavigator() {
+  return (
+    <Tab.Navigator
+      tabBarOptions={{
+        activeTintColor: Colors.activeTintColor,
+        inactiveTintColor: Colors.inactiveTintColor,
+        safeAreaInsets: {
+          top: 5,
```
Comment (satya164) [inline apps/bare-expo/MainNavigator.tsx:95]: I think top safe area inset doesn't do anything in bottom tab navigator.

File: apps/bare-expo/MainNavigator.tsx [hunk 5]
```diff
@@ -53,40 +43,76 @@ const Redirect = optionalRequire(() =>
...
-      top: 5,
-      right: 'always',
-      bottom: 'always',
-      left: 'always',
+const Tab = createBottomTabNavigator();
+
+const Switch = createStackNavigator();
+
+const linking = {
+  prefixes: [Platform.select({ web: Linking.makeUrl('/'), default: 'bareexpo://' })],
```
Comment (satya164) [inline apps/bare-expo/MainNavigator.tsx:55]: You can leave out the conditional. The `prefixes` option is ignored on web.

File: apps/bare-expo/MainNavigator.tsx [hunk 6]
```diff
@@ -53,40 +43,76 @@ const Redirect = optionalRequire(() =>
...
+        inactiveTintColor: Colors.inactiveTintColor,
+        safeAreaInsets: {
+          top: 5,
+        },
+      }}
+      initialRouteName="test-suite">
+      {Object.keys(routes).map(name => (
+        <Tab.Screen
+          name={name}
+          key={name}
```
Comment (satya164) [inline apps/bare-expo/MainNavigator.tsx:102]: Technically `key` is not needed here, but I guess ESLint complains 😛

File: apps/native-component-list/App.tsx [hunk 7]
```diff
@@ -2,27 +2,36 @@ import { AppLoading } from 'expo';
...
-export default function AppContainer() {
-  return (
-    <SafeAreaProvider>
-      <AppearanceProvider>
-        <App />
-      </AppearanceProvider>
-    </SafeAreaProvider>
-  );
+if (Platform.OS === 'android') {
+  enableScreens(true);
```
Comment (satya164) [inline apps/native-component-list/App.tsx:31]: Currently React Navigation only uses screens on Android so the conditional is nut needed if you're not using screens elsewhere.

File: apps/native-component-list/src/navigation/RootNavigation.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+import RedirectScreen from '../screens/RedirectScreen';
+import MainTabNavigator from './MainTabNavigator';
+
+const Switch = createStackNavigator();
+
+export const linking = {
+  prefixes: [Linking.makeUrl('/')],
+  config: {
+    main: {
```
Comment (satya164) [inline apps/native-component-list/src/navigation/RootNavigation.tsx:14]: same comments as `MainNavigator`

File: apps/native-component-list/src/screens/Image/ImageTestScreen.tsx [hunk 9]
```diff
@@ -1,179 +1,143 @@
...
 const AnimatedImage = Animated.Image;
 AnimatedImage.displayName = 'Image';
 
-type StateType = {
-  animValue?: Animated.Value;
-  viewKey: string;
-  events: string[];
+let compareEnabled: boolean = false;
+
+type Props = {
```
Comment (satya164) [inline apps/native-component-list/src/screens/Image/ImageTestScreen.tsx:26]: You can also do:

```ts
type Props = StackScreenProps<Links, 'ImageTest'>;
```

File: apps/test-suite/App.js [hunk 10]
```diff
@@ -1,7 +1,18 @@
...
 
 import AppNavigator from './AppNavigator';
 
-const App = createAppContainer(AppNavigator);
-
-export default App;
+const linking = {
+  prefixes: [Linking.makeUrl('/')],
+  config: {
+    select: '',
```
Comment (satya164) [inline apps/test-suite/App.js:10]: Same comments regarding linking as before

**Pull request merged**